src/core/lib/event_engine/thread_pool/thread_count.h

```diff
@@ -0,0 +1,95 @@
+// Copyright 2023 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#include <grpc/support/port_platform.h>
+
+#include <atomic>
+#include <cstddef>
+#include <vector>
+
+#include "absl/base/thread_annotations.h"
+
+#include "src/core/lib/gprpp/sync.h"
+#include "src/core/lib/gprpp/time.h"
+
+namespace grpc_event_engine {
+namespace experimental {
+
+// Tracks counts across some fixed number of shards.
+// It is intended for fast increment/decrement operations, but a slower overall
+// count operation.
+class BusyThreadCount {
+ public:
+  // Increments a per-shard counter on construction, decrements on destruction.
+  class AutoThreadCounter {
```

Comment:
We should implement move constructors or mark them deleted here - it's not immediately clear that MakeAutoThreadCounter is sound.

Reply:
so move only would be completely valid, as would unmovable/uncopyable (although that might make `MakeAutoThreadCounter` not work?).

Move only needs an `if` in the destructor to deal with destroying a now uncounted object, unmovable does not - the compiler ought to be able to optimize that away though (in most cases).

Reply:
I see what you're saying now, we're looking at a double-decrement if we don't fix the moved-from type.

Reply:
Fixed.